python/delta_sharing/converter.py

```diff
@@ -14,44 +14,106 @@
+    data types.
+    :param schema_json: json object representing the table schema
+    :return: empty table with columns specified in schema_json
+    """
+    assert schema_json["type"] == "struct"
+
+    dummy_table = pd.DataFrame(
+        {field["name"]: _get_dummy_column(field["type"]) for field in schema_json["fields"]}
+    )
+    return dummy_table.iloc[0:0]
+
+
+def to_converters(schema_json: dict) -> Dict[str, Callable[[str], Any]]:
     assert schema_json["type"] == "struct"
 
     return {field["name"]: to_converter(field["type"]) for field in schema_json["fields"]}
 
 
-def to_converter(json) -> Callable[[str], Any]:
-    if json == "boolean":
+def to_converter(schema_type) -> Callable[[str], Any]:
+    """
+    For types that support partitioning, a lambda to parse data into the
+    corresponding type is returned. For data types that cannot be partitioned
+    on, we return None. The caller is expected to check if the value is None before using.
+    :param schema_type: str or json representing a data type
+    :return: converter function or None
+    """
+    if schema_type == "boolean":
         return lambda x: None if (x is None or x == "") else (x is True or x == "true")
-    elif json == "byte":
+    elif schema_type == "byte":
         return lambda x: np.nan if (x is None or x == "") else np.int8(x)
-    elif json == "short":
+    elif schema_type == "short":
         return lambda x: np.nan if (x is None or x == "") else np.int16(x)
-    elif json == "integer":
+    elif schema_type == "integer":
         return lambda x: np.nan if (x is None or x == "") else np.int32(x)
-    elif json == "long":
+    elif schema_type == "long":
         return lambda x: np.nan if (x is None or x == "") else np.int64(x)
-    elif json == "float":
+    elif schema_type == "float":
         return lambda x: np.nan if (x is None or x == "") else np.float32(x)
-    elif json == "double":
+    elif schema_type == "double":
         return lambda x: np.nan if (x is None or x == "") else np.float64(x)
-    elif isinstance(json, str) and json.startswith("decimal"):
+    elif isinstance(schema_type, str) and schema_type.startswith("decimal"):
         return lambda x: None if (x is None or x == "") else Decimal(x)
-    elif json == "string":
+    elif schema_type == "string":
         return lambda x: None if (x is None or x == "") else str(x)
-    elif json == "date":
+    elif schema_type == "date":
         return lambda x: None if (x is None or x == "") else pd.Timestamp(x).date()
-    elif json == "timestamp":
+    elif schema_type == "timestamp":
         return lambda x: pd.NaT if (x is None or x == "") else pd.Timestamp(x)
+    elif schema_type == "binary":
+        return None  # partition on binary column not supported
+    elif isinstance(schema_type, dict) and schema_type["type"] in ("array", "struct", "map"):
+        return None  # partition on complex column not supported
 
     # TODO: binary
```

Comment:
Removing these will actually break a lot of other stuff since other parts of the code use the keys of `converters` to iterate over the columns of the table. We change things to use `schema_json`, but it'll be a lot of work and Ryan said it wasn't worth it.

Reply:
I mean the comment `# TODO: binary`.